Indicators/ImpliedVolatility.cs

```diff
@@ -241,33 +446,56 @@ protected override decimal ComputeNextValue(IndicatorDataPoint input)
+                case OptionPricingModelType.ForwardTree:
+                    return OptionGreekIndicatorsHelper.ForwardTreeTheoreticalPrice(volatility, spotPrice, strikePrice, timeToExpiration, riskFreeRate, dividendYield, optionType);
                 case OptionPricingModelType.BlackScholes:
                 default:
                     return OptionGreekIndicatorsHelper.BlackTheoreticalPrice(volatility, spotPrice, strikePrice, timeToExpiration, riskFreeRate, dividendYield, optionType);
             }
         }
 
-        // Calculate the IV of the option
-        private decimal CalculateIV(DateTime time)
+        /// <summary>
+        /// Computes the IV of the option
+        /// </summary>
+        /// <param name="time">the datetime at this calculation</param>
+        /// <returns>Smoothened IV of the option</returns>
+        protected virtual decimal CalculateIV(DateTime time)
         {
             RiskFreeRate.Update(time, _riskFreeInterestRateModel.GetInterestRate(time));
             DividendYield.Update(time, _dividendYieldModel.GetDividendYield(time));
 
-            var price = Price.Current.Value;
-            var spotPrice = UnderlyingPrice.Current.Value;
             var timeToExpiration = Convert.ToDecimal((Expiry - time).TotalDays) / 365m;
 
-            Func<double, double> f = (vol) => (double)(price - TheoreticalPrice(
-                Convert.ToDecimal(vol), spotPrice, Strike, timeToExpiration, RiskFreeRate.Current.Value, DividendYield.Current.Value, Right, _optionModel));
+            var impliedVol = 0m;
             try
             {
-                return Convert.ToDecimal(Brent.FindRoot(f, 1e-7d, 2.0d, 1e-4d, 100));
+                Func<double, double> f = (vol) => (double)(TheoreticalPrice(
+                    Convert.ToDecimal(vol), UnderlyingPrice, Strike, timeToExpiration, RiskFreeRate, DividendYield, Right, _optionModel) - Price);
+                impliedVol = Convert.ToDecimal(Brent.FindRoot(f, 1e-7d, 2.0d, 1e-4d, 100));
             }
             catch
             {
                 Log.Error("ImpliedVolatility.CalculateIV(): Fail to converge, returning 0.");
-                return 0m;
             }
+
+            if (UseMirrorContract)
+            {
+                var mirrorImpliedVol = 0m;
+                try
+                {
+                    Func<double, double> f = (vol) => (double)(TheoreticalPrice(
+                        Convert.ToDecimal(vol), UnderlyingPrice, Strike, timeToExpiration, RiskFreeRate, DividendYield, _oppositeOptionSymbol.ID.OptionRight, _optionModel) - _oppositePrice);
+                    mirrorImpliedVol = Convert.ToDecimal(Brent.FindRoot(f, 1e-7d, 2.0d, 1e-4d, 100));
+                }
+                catch
+                {
+                    Log.Error("ImpliedVolatility.CalculateIV(): Fail to converge, returning 0.");
+                }
```

Comment:
I think the whole try catch can be moved into a method passing in the option right/price to avoid the duplication. Can create another method for `Func<double, double> f = (vol) => (double)(TheoreticalPrice(` so it's not created every time